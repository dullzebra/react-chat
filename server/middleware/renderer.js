import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { SheetsRegistry } from 'react-jss/lib/jss';
import JssProvider from 'react-jss/lib/JssProvider';
import { createGenerateClassName } from 'material-ui/styles';

import configureStore from '../../src/store';
import App from '../../src/components/App';

const sheetsRegistry = new SheetsRegistry();
const generateClassName = createGenerateClassName();

const path = require('path');
const fs = require('fs');

export default (req, res) => {
  // point to the html file created by CRA's build tool
  const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');
  const context = {};
  const store = configureStore();

  fs.readFile(filePath, 'utf8', (err, htmlData) => {
    if (err) {
      console.error('err', err);
      return res.status(404).end();
    }

    if (context.url) {
      res.redirect(context.url);
      return;
    }

    // render the app as a string
    /* eslint-disable-next-line */
    const html = ReactDOMServer.renderToString(
      <JssProvider registry={sheetsRegistry} generateClassName={generateClassName}>
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            <App />
          </StaticRouter>
        </Provider>
      </JssProvider>);

    const css = sheetsRegistry.toString();

    // inject the rendered app into our html and send it
    return res.end(htmlData.replace(
      '<div id="root"></div>',
      `<div id="root">${html}</div>
    <script>window.INITIAL_STATE = ${JSON.stringify(store.getState()).replace(
    /</g,
    '\\u003c',
  )}</script>
    <style id="jss-server-side">${css}</style>
    `,
    ));
  });
};
