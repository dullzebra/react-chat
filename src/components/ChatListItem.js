import React from 'react';
import moment from 'moment';
import { withStyles } from 'material-ui/styles';
import PropTypes from 'prop-types';
import { ListItem, ListItemText } from 'material-ui/List';
import { NavLink } from 'react-router-dom';
import ChatAvatar from './ChatAvatar';

const styles = () => ({
  currentChat: {
    background: 'lightGrey',
  },
});

const ChatListItem = ({
  classes, disabled, id, title, date,
}) => (
  <ListItem
    button
    disabled={disabled}
    component={NavLink}
    to={`/chat/${id}`}
    activeClassName={classes.currentChat}
  >
    <ChatAvatar name={title} />
    <ListItemText primary={title} secondary={moment(date).fromNow()} />
  </ListItem>
);

ChatListItem.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  disabled: PropTypes.bool.isRequired,
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
};

export default withStyles(styles)(ChatListItem);
