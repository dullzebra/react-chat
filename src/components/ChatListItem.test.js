/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ChatListItem from './ChatListItem';

jest.mock('moment', () => () => ({ fromNow: () => '3 дня назад' }));

const mockData = {
  id: 'abc',
  title: 'Meoooow',
  date: '2018-03-23T08:06:56.789Z',
  disabled: false,
};

describe('<ChatListItem />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <ChatListItem {...mockData} />
      </MemoryRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<MemoryRouter>
        <ChatListItem {...mockData} />
              </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
