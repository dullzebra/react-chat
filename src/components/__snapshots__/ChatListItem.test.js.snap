// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatListItem /> renders correctly 1`] = `
<a
  aria-current={false}
  className="MuiButtonBase-root-65 MuiListItem-root-55 MuiListItem-default-58 MuiListItem-gutters-62 MuiListItem-button-63"
  href="/chat/abc"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  style={undefined}
  tabIndex={0}
>
  <div
    className="MuiAvatar-root-67 MuiAvatar-colorDefault-68"
    style={
      Object {
        "backgroundColor": "#ff5722",
      }
    }
  >
    M
  </div>
  <div
    className="MuiListItemText-root-70"
  >
    <h3
      className="MuiTypography-root-76 MuiTypography-subheading-83 MuiListItemText-primary-73"
    >
      Meoooow
    </h3>
    <p
      className="MuiTypography-root-76 MuiTypography-body1-85 MuiTypography-colorTextSecondary-98 MuiListItemText-secondary-74"
    >
      3 дня назад
    </p>
  </div>
  <span
    className="MuiTouchRipple-root-100"
  />
</a>
`;
